scripts/magpietts/update_magpie_shars.py

```diff
@@ -0,0 +1,245 @@
+            torch.distributed.all_gather_object(gathered, (cuts_local, codes_local, context_codes_local))
+        else:
+            gathered = [(cuts_local, codes_local, context_codes_local)]
+
+        if trainer.global_rank == 0:
+            for cuts_rank, codes_rank, context_codes_rank in gathered:
+                for cut, code, context_code in zip(cuts_rank, codes_rank, context_codes_rank):
+                    cut = cut.attach_tensor(
+                        name=f"codes_{self.codec_model_name}",
+                        data=code.cpu().type(torch.int16).numpy(),
+                    ).attach_tensor(
+                        name=f"context_codes_{self.codec_model_name}",
+                        data=context_code.cpu().type(torch.int16).numpy(),
+                    )
+                    self.cuts_buffer.append(cut)
+
+            if len(self.cuts_buffer) >= self.buffer_size:
+                self._write_buffer()
+            
+        et = time.time()
+        print("Time to write:", et - st)
+
+    def _write_buffer(self):
+        print("Writing buffer of size:", len(self.cuts_buffer))
+        for cut in self.cuts_buffer:
+            # print("Writing cut:", cut.id)
+            processed_cut = self.process_cut_for_saving(cut)
+            self.shar_writer.write(processed_cut)
+        self.cuts_buffer.clear()
+
+    def teardown(self, trainer, pl_module, stage: str | None = None):
+        if trainer.world_size > 1:
+            torch.distributed.barrier()
+
+        if trainer.global_rank == 0 and self.shar_writer is not None:
+            if self.cuts_buffer:
+                self._write_buffer()
+            self.shar_writer.close()
+            self.shar_writer = None
+
+def ddp_info():
+    if torch.distributed.is_available() and torch.distributed.is_initialized():
+        return torch.distributed.get_world_size(), torch.distributed.get_rank()
+    return 1, 0
+
+class CodecExtractor(pl.LightningModule):
+    def __init__(self, model_path, pad_multiple, batch_size, shar_root, num_workers):
+        super().__init__()
+        self.pad_multiple = pad_multiple
+        self.codec_model = AudioCodecModel.restore_from(restore_path=model_path, strict=False)
+        self.codec_model.eval()
+        self._dataloader = None
+        self.batch_size = batch_size
+        self.shar_root = shar_root
+        self.num_workers = num_workers
+        self.total_cuts = 0
+
+    def setup(self, stage=None):
+        if self._dataloader is None:
+            world_size, rank = ddp_info()
+            print("In model.setup - world size:", world_size, "rank:", rank)
+            cuts = CutSet.from_shar(in_dir=self.shar_root)
+            self.total_cuts = len(cuts)
+            self.approx_total_batches = self.total_cuts // (self.batch_size * world_size)
+            sampler = SimpleCutSampler(
+                cuts, shuffle=False, max_cuts=self.batch_size,
+                world_size=world_size, rank=rank,
+            )
```

Comment:
this may shard different items from the origin cuts shard. I observed this issue from my former attempts. it is worth to be double-checked. 

FYI, I implemented v2 version of lhotse shar prep, and avoided this issue. Will post a PR here once I got feedbacks from Piotr.

Brief intro: now the whole lhotse shar recipe was split into two steps: nemo manifest -> cuts/target_audio/context_audio shards -> extend with target_codes/context_codes.
Step 1: https://gitlab-master.nvidia.com/xueyang/nemo/-/blob/xueyang/magpie_release_2504-lhotse-v2-debug/scripts/magpietts/convert_nemo_to_lhotse_shar.py?ref_type=heads
https://gitlab-master.nvidia.com/xueyang/nemo/-/blob/xueyang/magpie_release_2504-lhotse-v2-debug/scripts/magpietts/extend_lhotse_shards_with_audio_codes.py?ref_type=heads

Reply:
I checked and the order of the cuts is preserved in the dataloader - Shuffle false and a fixed batch size should be the reason.